feat: add custom signature tag registry options

mrjoelkamp opened the pull request:
<!--
Thanks for opening a pull request!

Please remember to:
- mention any issue(s) that this PR closes using a closing keyword as well as the issue number, such as "Closes #XYZ" or "Resolves sigstore/repo-name#XYZ", cf.
  [documentation](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)
- ensure your commits are signed-off, as sigstore uses the [DCO](https://en.wikipedia.org/wiki/Developer_Certificate_of_Origin) using `git commit -s`, or `git commit -s --amend` if you want to amend already existing commits
- lastly, ensure there are no merge commits!
Thank you :)
-->

#### Summary
<!--
A description of what this pull request does, as well as QA test steps (if applicable and if not already added to the Jira ticket).
-->
Added `TagPrefix` and `TagSuffix` to `TagOptions` and added `TagOptions` to `RegistryOptions` to pass custom tags to `ociremote.Options` to allow customization of the signature tag where:
signature tag = `[OptionalTagPrefix]sha256-[Digest].[TagExtension]`

And the CLI flags `--tag-prefix` and `--tag-suffix` to set these options 

#### Ticket Link
<!--
If this pull request addresses a Help Wanted ticket, please link the relevant GitHub issue, e.g.

  Fixes https://github.com/sigstore/YYYYYY/issues/XXXXX

-->
Fixes
https://github.com/sigstore/cosign/issues/807

cc: @n3wscott 

#### Release Note
<!--
Add a release note for each of the following conditions:

* Config changes (additions, deletions, updates)
* API additions—new endpoint, new response fields, or newly accepted request parameters
* Database changes (any)
* Websocket additions or changes
* Anything noteworthy to a Mattermost instance administrator (err on the side of over-communicating)
* New features and improvements, including behavioural changes, UI changes and CLI changes
* Bug fixes and fixes of previous known issues
* Deprecation warnings, breaking changes, or compatibility notes

If no release notes are required write NONE. Use past-tense.

-->
```release-note
NONE
```

mattmoor commented on cmd/cosign/cli/options/flags.go:
```diff
@@ -44,6 +44,8 @@ func NOf(args ...interface{}) int {
 func ApplyRegistryFlags(regOpts *RegistryOptions, fs *flag.FlagSet) {
 	fs.BoolVar(&regOpts.AllowInsecure, "allow-insecure-registry", false, "whether to allow insecure connections to registries. Don't use this for anything but testing")
+	fs.StringVar(&regOpts.TagPrefix, "signature-prefix", "", "custom prefix to use for signature tag")
+	fs.StringVar(&regOpts.TagSuffix, "signature-suffix", "", "custom suffix to use for signature tag")
```
Hmm, these don't really feel like registry flags to me.

I think surfacing a way to customize suffix has come up a few times now, and this totally makes sense.  I think we should probably plumb through a similar change for attestations.

Especially if we leave the field in such a generic place, we should have the field name's specificity match the flag.  I could see `--suffix` setting `TagSuffix` or `--signature-suffix` setting `SignatureTagSuffix`, but as it is seems a bit odd.

mattmoor replied:
cc @dekkagaijin @n3wscott  for thoughts on where this lives, and consistency with other flags.

mrjoelkamp replied:
@mattmoor I agree, but the exiting suffix code was living in registry options so I extended it to include prefix and exposed the existing `TagSuffix` as a flag as well. `TagPrefix` is more important to me because of ECR Lifecycle Policy tag prefix filtering.

I'll push an update to make the `TagPrefix` signature, SBOM, and attestation agnostic.

mattmoor commented on cmd/cosign/cli/options/registry.go:
```diff
@@ -36,10 +38,16 @@ var _ Interface = (*RegistryOptions)(nil)
 
 func (o *RegistryOptions) ClientOpts(ctx context.Context) []ociremote.Option {
-	return []ociremote.Option{ociremote.WithRemoteOptions(o.GetRegistryClientOpts(ctx)...)}
+	if o.TagSuffix != "" {
+		return []ociremote.Option{ociremote.WithSignatureSuffix(o.TagSuffix), ociremote.WithSignaturePrefix(o.TagPrefix), ociremote.WithRemoteOptions(o.GetRegistryClientOpts(ctx)...)}
+	} else {
+		return []ociremote.Option{ociremote.WithSignaturePrefix(o.TagPrefix), ociremote.WithRemoteOptions(o.GetRegistryClientOpts(ctx)...)}
+	}
```
I guess I found it odd that this passes `TagPrefix` without a similar `""` check, which sort of bakes in the idea that `""` is the default.  Perhaps we should build this up?

mrjoelkamp replied:
@mattmoor, I needed to conditionally check for `TagSuffix` because it would override `SignatureTagSuffix` to `""` instead of the default `.sig`. There wasn't a `TagPrefix` before so I defaulted it to `""` and don't need to perform the same check

mattmoor commented on pkg/oci/remote/options.go:
```diff
@@ -78,6 +81,14 @@ func makeOptions(target name.Repository, opts ...Option) (*options, error) {
 	return o, nil
 }
 
+// WithSignaturePrefix is a functional option for overriding the default
+// signature tag prefix.
+func WithSignaturePrefix(prefix string) Option {
+	return func(o *options) {
+		o.SignaturePrefix = prefix
+	}
+}
+
```
A lot of the plumbing seems to be adding the prefix support, which I don't fully understand the need for, if folks can customize the suffix.  Seems like a lot of knobs, where just suffix would hopefully suffice.

mrjoelkamp replied:
@mattmoor I started with suffix only, until I realized that ECR Lifecycle Policies only allow for [tag prefixes](https://docs.aws.amazon.com/AmazonECR/latest/userguide/LifecyclePolicies.html#lp_tag_prefix_list).

I added the prefix option to account for that and left in the suffix option in case someone wanted to use something different than `.sig`. I'd be okay will pulling the suffix code out but really need the prefix code at the moment.

mattmoor replied:
`/me shakes fist at ECR`

I think I'd prefer we nail this for suffix (incl. stuff like `.att`) and then mirror everything for prefix.  Sorry if that's annoying.

mrjoelkamp replied:
@mattmoor it sounds like this PR brought attention to a bigger code organization issue. I only exposed the existing `WithSignatureSuffix` to the CLI flags and added `WithPrefix` to the existing code base. Any chance we could merge this to get the feature in and then refactor how tags are formed in a separate PR?

samj1912 commented:
Would it be better to introduce a gotemplate flag or something along those lines instead? It could be passed certain values like registry, digest, tag, image name etc and it would be left upto the user to have consistent templates during sign and verify?

dekkagaijin commented:
imho `[OptionalTagPrefix]sha256-[Digest].[TagExtension]` would be preferable, since it allows for "dumb"er implementation and keeps the conceptually atomic `sha256-[Digest]` in one piece.

We might consider allowing overriding `fn(image) -> attachment_tag` in the library, but for the CLI lets stick with `--attachment-tag-prefix`

mrjoelkamp commented:
> imho `[OptionalTagPrefix]sha256-[Digest].[TagExtension]` would be preferable, since it allows for "dumb"er implementation and keeps the conceptually atomic `sha256-[Digest]` in one piece.
> 
> We might consider allowing overriding `fn(image) -> attachment_tag` in the library, but for the CLI lets stick with `--attachment-tag-prefix`

@dekkagaijin I refactored a few things and I think that I have addressed most of the comments! How does it look?

dekkagaijin commented on cmd/cosign/cli/options/tags.go:
```diff
@@ -0,0 +1,34 @@
+var _ Interface = (*TagOptions)(nil)
+
+// AddFlags implements Interface
+func (o *TagOptions) AddFlags(cmd *cobra.Command) {
+	cmd.Flags().StringVar(&o.TagPrefix, "tag-prefix", "", "custom prefix to use for tags")
+	cmd.Flags().StringVar(&o.TagSuffix, "tag-suffix", "", "custom suffix to use for tags")
+}
```
@mattmoor wdty about standardizing on "attachment image tag suffix" == "attachment name"

mrjoelkamp commented:
> @mrjoelkamp If you resolve the docgen issue(`go run ./cmd/help/`), I'm happy to merge for now to save you more conflict grief

Awesome, thanks! Should be good to go now.
